File: python/utils/polling.py
"""
Polling utility for async video generation

Use this to wait for video generation to complete.
Status checks are FREE - no payment required!
"""

import time
import requests
from typing import Callable, Optional, Dict, Any


def poll_status(
    status_url: str,
    interval: int = 5,
    max_attempts: int = 60,
    on_status: Optional[Callable[[Dict[Any, Any], int, int], None]] = None
) -> Dict[Any, Any]:
    """
    Poll a status endpoint until completion
    
    Args:
        status_url: Full URL to status endpoint
        interval: Polling interval in seconds (default: 5)
        max_attempts: Maximum attempts (default: 60)
        on_status: Callback for status updates
        
    Returns:
        Final status data
        
    Raises:
        Exception: If generation fails or polling times out
    """
    attempt = 0
    
    while attempt < max_attempts:
        time.sleep(interval)
        attempt += 1
        
        response = requests.get(status_url)
        data = response.json()
        
        # Call status callback if provided
        if on_status:
            on_status(data, attempt, max_attempts)
        
        # Check if completed
        if data['status'] == 'completed':
            return data
        
        # Check if failed
        if data['status'] == 'failed':
            error_msg = data.get('error', 'Unknown error')
            raise Exception(f'Generation failed: {error_msg}')
        
        # Continue polling for: pending, in_queue, in_progress
    
    raise Exception(f'Polling timeout after {max_attempts} attempts')


def wait_for_video(
    generation_id: str,
    base_url: str,
    interval: int = 5,
    max_attempts: int = 60,
    on_status: Optional[Callable[[Dict[Any, Any], int, int], None]] = None
) -> Dict[Any, Any]:
    """
    Wait for video generation to complete
    
    Args:
        generation_id: Generation ID from generate response
        base_url: API base URL
        interval: Polling interval in seconds (default: 5)
        max_attempts: Maximum attempts (default: 60)
        on_status: Optional custom status callback
        
    Returns:
        Final video data with URL
    """
    status_url = f"{base_url}/v1/status/{generation_id}"
    
    print('⏳ Waiting for video generation...')
    
    def default_on_status(data: Dict[Any, Any], attempt: int, max_attempts: int):
        status_emoji = {
            'pending': '⏸️',
            'in_queue': '🔄',
            'in_progress': '🎬',
            'completed': '✅',
            'failed': '❌'
        }.get(data['status'], '❓')
        
        print(f"{status_emoji} [{attempt}/{max_attempts}] Status: {data['status']}")
        
        # Call custom callback if provided
        if on_status:
            on_status(data, attempt, max_attempts)
    
    return poll_status(
        status_url,
        interval=interval,
        max_attempts=max_attempts,
        on_status=default_on_status
    )

File: javascript/utils/polling.js
/**
 * Polling utility for async video generation
 * 
 * Use this to wait for video generation to complete.
 * Status checks are FREE - no payment required!
 */

/**
 * Poll a status endpoint until completion
 * @param {string} statusUrl - Full URL to status endpoint
 * @param {object} options - Polling options
 * @param {number} options.interval - Polling interval in ms (default: 5000)
 * @param {number} options.maxAttempts - Maximum attempts (default: 60)
 * @param {function} options.onStatus - Callback for status updates
 * @returns {Promise<object>} Final status data
 */
export async function pollStatus(statusUrl, options = {}) {
  const {
    interval = 5000,
    maxAttempts = 60,
    onStatus = null
  } = options;
  
  let attempt = 0;
  
  while (attempt < maxAttempts) {
    await new Promise(resolve => setTimeout(resolve, interval));
    attempt++;
    
    const response = await fetch(statusUrl);
    const data = await response.json();
    
    // Call status callback if provided
    if (onStatus) {
      onStatus(data, attempt, maxAttempts);
    }
    
    // Check if completed
    if (data.status === 'completed') {
      return data;
    }
    
    // Check if failed
    if (data.status === 'failed') {
      throw new Error('Generation failed: ' + (data.error || 'Unknown error'));
    }
    
    // Continue polling for: pending, in_queue, in_progress
  }
  
  throw new Error(`Polling timeout after ${maxAttempts} attempts`);
}

/**
 * Wait for video generation to complete
 * @param {string} generationId - Generation ID from generate response
 * @param {string} baseUrl - API base URL
 * @param {object} options - Polling options
 * @returns {Promise<object>} Final video data with URL
 */
export async function waitForVideo(generationId, baseUrl, options = {}) {
  const statusUrl = `${baseUrl}/v1/status/${generationId}`;
  
  console.log('⏳ Waiting for video generation...');
  
  return await pollStatus(statusUrl, {
    ...options,
    onStatus: (data, attempt, maxAttempts) => {
      const statusEmoji = {
        'pending': '⏸️',
        'in_queue': '🔄',
        'in_progress': '🎬',
        'completed': '✅',
        'failed': '❌'
      }[data.status] || '❓';
      
      console.log(`${statusEmoji} [${attempt}/${maxAttempts}] Status: ${data.status}`);
      
      // Call custom callback if provided
      if (options.onStatus) {
        options.onStatus(data, attempt, maxAttempts);
      }
    }
  });
}

